src/subdataframe/subdataframe.jl

```diff
@@ -223,3 +298,55 @@ function DataFrame(sdf::SubDataFrame; copycols::Bool=true)
 end
 
 Base.convert(::Type{DataFrame}, sdf::SubDataFrame) = DataFrame(sdf)
+
+# this function tests if it is allowed to add columns to passed SubDataFrame
+# currently it is only allowed when SubDataFrame was created with : as column selector
+# which results in using Index as its index (as opposed to other columns selectors
+# which result in SubIndex)
+function is_column_insertion_allowed(df::AbstractDataFrame)
+    if df isa DataFrame
+        return true
+    elseif df isa SubDataFrame
+        return getfield(df, :colindex) isa Index
+    end
+    throw(ArgumentError("Unsupported data frame type"))
+end
+
+function _replace_columns!(sdf::SubDataFrame, newdf::DataFrame)
+    colsmatch = _names(sdf) == _names(newdf)
+
+    if !(colsmatch || is_column_insertion_allowed(sdf))
+        throw(ArgumentError("changing the sequence of column names in a SubDataFrame " *
+                            "that subsets columns of its parent data frame is disallowed"))
+    end
+
+    for colname in _names(newdf)
+        oldcol = sdf[!, colname]
+        newcol = newdf[!, colname]
+        # We perform an in-place operation if possible for performance.
+        # This has an additional effect that for CategoricalVector levels
+        # and ordering will be retained or not depending on which code patch is taken.
+
+        # TODO: add tests when promote_type vs Base.promote_typejoin decision is made
```

Comment:
@nalimilan - this is the performance optimization we have discussed for `select!` and `transform!`. As commented - it has a side effect for CategoricalArrays.jl but I think it is OK.

Reply:
Ah, but this isn't just an optimization, as the difference can be observed from the outside, right? I think we should only apply tricks that are completely invisible for users, or the behavior will be to complex. Apparently I was wrong in thinking that `select!` and `transform` are amenable to such optimizations.

Reply:
Yes - it is visible from outside if the column has metadata. I will restrict the optimization to `Vector`, as `Vector` has no metadata (as opposed to `PooledVector` or `CategoricalVector`)

Reply:
Isn't it visible also for `Vector`? That is, if one holds a reference to the column, mutating it rather than replacing it can have consequences. (I'm not overly concerned about `PooledVector` and `CategoricalVector` if that only makes a difference in weird cases.)

Reply:
It can have consequences if there are aliases to this vector. But OK - let us just copy always and stick to "safety first" rule (if someone wants it fast it is always easy-enough to achieve with indexing).